x-pack/plugins/infra/public/components/autocomplete_field/autocomplete_field.tsx

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  EuiFieldSearch,
+  EuiFieldSearchProps,
+  EuiOutsideClickDetector,
+  EuiPanel,
+} from '@elastic/eui';
+import React from 'react';
+import styled from 'styled-components';
+
+import { AutocompleteSuggestion } from 'ui/autocomplete_providers';
+
+import { composeStateUpdaters } from '../../utils/typed_react';
+import { SuggestionItem } from './suggestion_item';
+
+interface AutocompleteFieldProps {
+  isLoadingSuggestions: boolean;
+  isValid: boolean;
+  loadSuggestions: (value: string, cursorPosition: number, maxCount?: number) => void;
+  onSubmit?: (value: string) => void;
+  onChange?: (value: string) => void;
+  suggestions: AutocompleteSuggestion[];
+  value: string;
+}
+
+interface AutocompleteFieldState {
+  areSuggestionsVisible: boolean;
+  selectedIndex: number | null;
+}
+
+export class AutocompleteField extends React.Component<
```

Comment:
I chose this name explicitly because there is no dependency on kuery in here. It is designed to work with any autocompletion mechanism as long as it can provide the required properties.

I also did not name it something like `FilterBar`, because it is merely a text input. It might be embedded inside of a filter bar along with other criteria (e.g. date range picker).

Does that make sense?

Reply:
To expand on the previous comment, the kuery-specific autocompletion is handled by the `WithKueryAutocompletion` container... very creative name, I know :wink:

Reply:
It is logical, but it still means that to find the UI element everyone calls "Kuery bar" (or just "Query bar") I have to make the mental leap that this is an `AutocompleteField` combined `withKueryAutocompletion`. Will we ever have different autocompletion mechanisms, and is it worth preparing for that case now?

Reply:
This specific case probably isn't worth spending a lot of discussion time on, it's more a generic question, because these mental overheads tend to pile up.

Reply:
I'm confused... do you mean you would like us to stop separating concerns into presentational components and containers?

Reply:
No.